packages/zoe/src/contractFacet/zcfSeat.js

```diff
@@ -334,32 +240,158 @@ export const createSeatManager = (
-    const zcfSeat = makeZCFSeatInternal(proposal);
-    activeZCFSeats.init(zcfSeat, initialAllocation);
-    zcfSeatToSeatHandle.init(zcfSeat, seatHandle);
-    return zcfSeat;
-  };
 
   const replaceDurableWeakMapStore = (baggage, key) => {
     const mapStore = makeScalarBigWeakMapStore(key, { durable: true });
     baggage.set(key, mapStore);
     return mapStore;
   };
 
-  /** @type {DropAllReferences} */
-  const dropAllReferences = () => {
-    activeZCFSeats = replaceDurableWeakMapStore(zcfBaggage, 'activeZCFSeats');
-    zcfSeatToSeatHandle = replaceDurableWeakMapStore(
-      zcfBaggage,
-      'zcfSeatToSeatHandle',
-    );
+  const ZcfSeatManagerGuard = {
+    seatManager: M.interface('ZcfSeatManager', {
+      makeZCFSeat: M.call(
+        M.partial({
+          proposal: ProposalShape,
+          initialAllocation: AmountKeywordRecordShape,
+          seatHandle: SeatShape,
+          offerArgs: M.any(),
+        }),
+      ).returns(M.remotable('zcfSeat')),
+      reallocate: M.call(M.remotable('zcfSeat'), M.remotable('zcfSeat'))
+        .rest(M.arrayOf(M.remotable('zcfSeat')))
+        .returns(),
+      dropAllReferences: M.call().returns(),
+    }),
+    zcfMintReallocator: M.interface('MintReallocator', {
+      reallocate: M.call(SeatShape, AmountKeywordRecordShape).returns(),
+    }),
   };
 
-  return harden({
-    makeZCFSeat,
-    reallocate,
-    reallocateForZCFMint,
-    dropAllReferences,
-  });
+  const makeSeatManagerKit = vivifyFarClassKit(
+    zcfBaggage,
+    'ZcfSeatManager',
+    ZcfSeatManagerGuard,
+    initEmpty,
+    {
+      seatManager: {
+        makeZCFSeat({ proposal, initialAllocation, seatHandle }) {
+          const zcfSeat = makeZCFSeatInternal(proposal);
+          activeZCFSeats.init(zcfSeat, initialAllocation);
+          zcfSeatToSeatHandle.init(zcfSeat, seatHandle);
+          return zcfSeat;
+        },
+
+        reallocate(/** @type {ZCFSeat[]} */ ...seats) {
+          seats.forEach(assertActive);
+          seats.forEach(assertStagedAllocation);
+
+          // Ensure that rights are conserved overall.
+          const flattenAllocations = allocations =>
+            allocations.flatMap(Object.values);
+
+          const previousAllocations = seats.map(seat =>
+            seat.getCurrentAllocation(),
+          );
+          const previousAmounts = flattenAllocations(previousAllocations);
+
+          const newAllocations = seats.map(seat => seat.getStagedAllocation());
+          const newAmounts = flattenAllocations(newAllocations);
+
+          assertRightsConserved(previousAmounts, newAmounts);
+
+          // Ensure that offer safety holds.
+          seats.forEach(seat => {
+            isOfferSafe(seat.getProposal(), seat.getStagedAllocation()) ||
+              assert.fail(
+                X`Offer safety was violated by the proposed allocation: ${seat.getStagedAllocation()}. Proposal was ${seat.getProposal()}`,
+              );
+          });
```

Comment:
IMO there is no strong reason to prefer either `forEach` or `for of` over the other. That said, I mildly prefer `for of` for a few reasons:
   * easier to follow into the loop body in the debugger.
   * can use a break or return for easy early exit.

Since the only early exit here is a throw, I do not see any correctness difference.

I think I marginally find `for of` more readable, but it is really close.

@gibson042 , curious about your reasons?

Reply:
>  trouble extracting `Fail` from `assert`. (I've already upgraded my `node`.)

@Chris-Hibbert , Doesn't depend on node. Comes from `ses` in `endo`, and thus depended on what version of `ses` your `agoric-sdk` depends on.

@gibson042 if you rebase gibson-4567-notifier-package-durability onto current master and @Chris-Hibbert re-rebased this PR on that, then this problem should be fixed. Then we won't have to remember to fix this later ;)